### DISCUSSION TITLE
2021年11月7日直播问题提前收集

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/designable#120] [Show and tell]
话题：designable源码分析，扩展开发相关话题
平台：B站直播
时间：晚上7点半，7点20开始暖场，可以闲聊一些无关紧要的东西
地址：https://b23.tv/pHsLnx

各位有啥问题评论在这里，直播间统一回答

**Comment by xiaoxiansheng1** [comment]
源码看不懂咋办

**Reply by janryWang** [reply]
看直播就对了

**Reply by dahui4dev** [reply]
同类问题+1 如何更快速高效阅读源码？

**Reply by hlzzd666** [reply]
同看不懂

**Comment by lifeneedspassion** [comment]
可以扩展成 WebFlow 那种程度吗？

**Comment by Lewis-li-0** [comment]


**Reply by liuweiGL** [reply]
作者给你跳个舞吧。

**Comment by muuyao** [comment]
formily 和 designable 抽象能力都很强，代码分层设计厉害，如何培养这种领域抽象和代码设计能力？

**Comment by knight-peter** [comment]
针对formily UI组件库 和 designable里的自定义组件，能不能出详细点的文档。
对于我这种水平比较菜的人，想要顺利的完成自定义组件真的很难。但是又很想用formily这种方便未来扩展的工具来实现需求。

**Comment by wsxe9988** [comment]
是否可以讲下，怎样扩展出自由布局，

**Comment by 1-WEEK** [comment]
react-settings-form 内部组件如何自定义扩展，自定义组件在设置表单可能会有独有的业务属性

**Reply by YqxLzx** [reply]
同问 大佬解决了吗

**Comment by aidening** [comment]
大佬，这里能写界面优化建议不

**Comment by feilaoda** [comment]
1、想在Select组件上的数据源增加一个远程数据源，而不只是现在的静态数据，思路是什么？把值设置到enum上？
2、对组件的显示、隐藏属性，只想在渲染时生效，在配置时的设计界面不生效（即不隐藏），怎么实现？

**Comment by zhouping3** [comment]
一、功能
1.响应器配置：当前的模式是被动依赖，后续会支持主动依赖模式吗
2.三方组件注入：后续会支持引入第三方或者自定义组件吗
3.dn-aux-focus这个helpers按钮具体有什么用，使用了没发现特别之处
4.自由布局的扩展，比如引入react-grid-layout的话会有一些拖拽效果上的副作用
二、源码
1.为什么撸了一套拖拽事件引擎，没基于第三方的拖拽库做加强，比如用的比较多的reat-dnd等等
2.常规的一些操作(拖拽生成组件、修改配置)底层主要是哪些models配合处理，以及处理的流程

**Comment by fongfai** [comment]
designable有什么规划 会衍生出类似lowcode的搭建平台的能力吗

**Comment by wsxe9988** [comment]
还可以分享下 designable在阿里内部的落地经验

**Comment by kkodeok** [comment]
1. 想知道如果自己想内部接入自己的组件或者类似antd 这类组件库的时候，是否能有个例子？
2. 如果自己使用designable的时候想做出类似于formily的playground这种拖拽，应该从什么地方入手，比如是直接在designable源代码上进行修改，还是起一个新的项目然后导入designable的npm包呢，以及希望可以有个简单例子。
3. 能否讲解下formily中json-schema的和react node还有mock schema之间的转换方法和逻辑，从而可以对接入第三方组件在这个位置提供一些思路或者说官方建议的标准？
4. 如果接入类似于button这种组件的话，想与其他的组件进行联动例如触发form的提交，触发modal的显示，目前的推荐方法是？（目前似乎只支持被动依赖这个不知道有没有具体的实现方案？）

**Comment by semmywong** [comment]
1. formili 2.x有很多break change,那么2.x里是否可以兼容正常使用1.x的联动方案？
2. 目前渲染要怎么摆脱根节点一定是Form的限制(根节点不是Form，会直接报错)，也就是渲染场景如果不是表单的情况下，要怎么办？
3. 文档里看到forlimy2.0的表单设计器还在开发，年底会放出来，请问。formily2.0表单设计器与1.0差别大吗？
4. connnect转换组件为formily组件时，第三个参数是mapReadPretty，那么是不是如果在Field属性中设置为UI形态为阅读，那么这个组件就是以当前mapReadPretty传入的组件形式展示，是吗？
5. formily官网上写着是对表单的解决方案，但是其实我觉得这个可以应用的范围已经超出表单的范畴，这个后面是否会更改这种定位目标？因为我觉得formily完全可以提升为页面级、甚至系统级别的渲染、布局方案。

**Comment by RihuaChen** [comment]
1. 想要自定义左侧组件跟右侧配置面板有没有什么好的推荐方式？
2. 国际化是否可以简化一些呢，现在国际化需要按组件层级编写，与大部分国际化的操作方式不太一致
3. Grid布局响应式断点如何配置呢，现在配置整个页面会挂掉
4. 能否参与帮忙完善文档的工作呢，需要什么可以先列个roadmap之类的，这边可以帮忙写一些demo哈，这样大家更有利于高效开发
5. https://rihuachen.github.io/docs/scene/submit 这是现在刚刚在做的formily的文档，我的想法是做一些给初级工程师使用的文档，最简单的落地的业务场景的这种，不知道合不合适做到官方文档里面，我暂时做给我们内部使用的

**Reply by YqxLzx** [reply]
第一条请问有解决方案了吗 我也想自定义右侧面板

**Comment by nekic** [comment]
现在把节点拖到 Designable Workspace时都默认变成 其children, 如何实现其他属性也可拖入节点，如 Card 组件，其title 也可以是一个React.Element的，这种情况该如何在schema中描述以及如何实现呢

**Comment by xiangxiong** [comment]
组件支持后端api的动态绑定数据源?
如何在搭建的页面自定义自己JS判断逻辑？
页面是动态渲染json 还是 渲染组件源码?

**Comment by sstruct** [comment]
1. 渲染器必须使用 formily 吗
2. 近期有啥规划

**Comment by RihuaChen** [comment]
感谢大佬的直播，受益匪浅，辛苦辛苦，已经尽量在公司内部推广formily了~

**Comment by janryWang** [comment]
Sandbox没跑起来的原因是要传入所有externals的资源

```tsx
 <Sandbox
                    jsAssets={[
                      'https://unpkg.com/react/umd/react.production.min.js',
                      'https://unpkg.com/react-dom/umd/react-dom.production.min.js',
                      'https://unpkg.com/antd/dist/antd-with-locales.min.js',
                      './sandbox.bundle.js',
                    ]}
                  />
```

**Comment by kaoding** [comment]
大佬，听你录播说到icon是你老婆设计，是左边的icon吗？很漂亮的，她在iconfont上有链接吗？ @janryWang

**Reply by janryWang** [reply]
没有的哈，她就单纯兴趣

**Reply by kaoding** [reply]
比一般的设计师水平还高。

**Comment by zhoulijunFE** [comment]
学习了大佬的直播，受益匪浅。直播中有讲到 远程获取数据源方式： 配置响应器 --> 动作响应 --> 一段 $effect(() 函数方式。
我想在 ArrayTable表格实现远程获取列表， 但是没读懂配置完这个$effect(() 函数，具体动态数据如何加载，如何渲染到ArrayTable组件的，想大佬大概给讲一下思路？@janryWang
